balancer/rls/internal/keys/builder.go

```diff
@@ -204,17 +225,20 @@ func (m matcher) Equal(a matcher) bool {
 	return true
 }
 
-func (b builder) keys(md metadata.MD) KeyMap {
+func (b builder) buildHeaderKeys(md metadata.MD) map[string]string {
 	kvMap := make(map[string]string)
-	for _, m := range b.matchers {
+	if md == nil {
```

Comment:
`if len(md) == 0` is probably even better.

Also it seems this is an allocation in the RPC path, so I'd `return nil` and move the `make` down below this `if` to avoid any unnecessary allocations.

And if it's possible to avoid this allocation altogether, that would be even better.

Reply:
Switched to checking for non-zero length.

We cannot avoid this allocation as the caller of this method, `RLSKey()`,  will try to add `extra_keys` and `constant_keys` to the returned map. We *could* add the following block before every possible assignment in `RLSKey`

```golang
if kvMap == nil {
  kvMap = make(map[string]string)
}
```
but that would be a lot of boilerplate. Wdyt?

Reply:
Oops, missed this.  No, what you have is fine in that case.